File: Week_07/547/friend_circles.go
package main

func findCircleNum(M [][]int) int {
	n := len(M)
	res := 0

	visited := make([]bool, n)

	var dfs func(i int)
	dfs = func(i int) {
		for j := 0; j < n; j++ {
			if !visited[j] && M[i][j] == 1 {
				visited[j] = true
				dfs(j)
			}
		}
	}
	for i := 0; i < n; i++ {
		if !visited[i] {
			res++
			visited[i] = true
			dfs(i)
		}
	}
	return res
}

File: Week_07/547/friend_circles.py
from typing import List


class Solution:
    def findCircleNum(self, M: List[List[int]]) -> int:
        n = len(M)
        res = 0
        visited = [False] * n

        def dfs(i: int):
            for j in range(n):
                if not visited[j] and M[i][j]:
                    visited[j] = True
                    dfs(j)

        for i in range(n):
            if not visited[i]:
                res += 1
                dfs(i)
        return res
